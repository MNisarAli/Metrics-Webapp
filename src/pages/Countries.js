import { useEffect, useMemo, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate, useLocation } from 'react-router-dom';
import { getCountries } from '../redux/Country/countriesSlice';
import Dashboard from '../components/Dashboard';
import Wrapper from '../components/Wrapper';

const Countries = () => {
  // hooks for accessing routing and state management functionality.
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const location = useLocation();
  const searchParams = new URLSearchParams(location.search);

  // parse the query params to get the continent info.
  const continent = JSON.parse(searchParams.get('continent') || '{}');

  // destructure the continent info.
  const {
    region, map, population, name, noOfCountries,
  } = continent;

  // get countries and status from the store.
  const { countries, status } = useSelector((state) => state.countries);

  // state for search term.
  const [searchTerm, setSearchTerm] = useState('');

  // fetch countries based on the continent info.
  useEffect(() => {
    const fetchCountries = async () => {
      dispatch(getCountries({ region, name }));
    };

    // fetch countries only if they are not already present or if the region/name has changed.
    if (
      countries.length === 0
      || region !== countries[0].region
      || name !== countries[0].continent
    ) {
      fetchCountries();
    }
  }, [name, region, countries, dispatch]);

  // handler for clicking on a country.
  const handleCountryClick = (country) => {
    navigate(`/country?country=${JSON.stringify(country)}`);
  };

  // handler for updating the search term.
  const handleSearch = (event) => {
    setSearchTerm(event.target.value);
  };

  // memoized array of filtered countries based on the search term.
  const filteredCountries = useMemo(() => countries.filter(
    (country) => country.name.toLowerCase().includes(searchTerm.toLowerCase()),
  ), [countries, searchTerm]);

  // show loading message if countries are still loading.
  if (status === 'loading') {
    return <div>Loading...</div>;
  }

  return (
    <>
      <Dashboard
        map={map}
        name={name}
        population={population}
      />
      <section className="data-section">
        <article className="search-wrapper">
          <h3 className="section-title">
            {noOfCountries}
            <span> Countries</span>
          </h3>
          <form>
            <input
              type="text"
              placeholder="Search country by name"
              className="search-input"
              onChange={handleSearch}
            />
          </form>
        </article>
        <nav className="navbar">
          {filteredCountries.map((country) => (
            <a
              className="nav-item"
              key={country.id}
              href={country.path}
              onClick={(e) => {
                e.preventDefault();
                handleCountryClick(country);
              }}
            >
              <Wrapper wrapper={country} />
            </a>
          ))}
        </nav>
      </section>
    </>
  );
};

export default Countries;
