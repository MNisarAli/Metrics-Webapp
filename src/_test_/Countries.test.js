import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import Countries from '../pages/Countries';
import store from '../Redux/configureStore';

describe('Countries page', () => {
  test('renders correctly', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <Countries />
        </BrowserRouter>
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
