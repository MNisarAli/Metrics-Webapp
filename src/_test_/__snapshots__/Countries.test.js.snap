// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countries page renders correctly 1`] = `
Array [
  <section
    className="dashboard"
  >
    <img
      alt=""
      className="dashboard-map"
      src=""
    />
    <article
      className="title-container"
    >
      <h2
        className="page-title"
      />
      <span />
    </article>
  </section>,
  <section
    className="data-section"
  >
    <article
      className="search-wrapper"
    >
      <h3
        className="section-title"
      >
        <span>
           Countries
        </span>
      </h3>
      <form>
        <input
          className="search-input"
          onChange={[Function]}
          placeholder="Search country by name"
          type="text"
        />
      </form>
    </article>
    <nav
      className="navbar"
    />
  </section>,
]
`;
